### PR TITLE
Add badges for organizations and datasets, fix #54

### LINKED ISSUE #54: Add badges for organizations and datasets
As an administrator, I want to be able to set/unset badges to organizations and datasets in order to highlight trusted sources.

### PR DESCRIPTION
**Pull request by davidbgk**

### CODE REVIEW

File: udata/core/badges/models.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+class Badge(db.EmbeddedDocument):
+    kind = db.StringField(choices=[], required=True)
+    created = db.DateTimeField(default=datetime.now, required=True)
+    created_by = db.ReferenceField('User')
+    removed = db.DateTimeField()
+    removed_by = db.ReferenceField('User')
+
+    meta = {
+        'allow_inheritance': True,
+        'ordering': ['created'],
```
Comment (noirbizarre) [inline udata/core/badges/models.py]: Maybe better with ``-created`` (last created first)

File: udata/core/dataset/api.py [hunk 2]
```diff
@@ -122,6 +126,39 @@ def delete(self, dataset):
         return dataset
 
 
+@ns.route('/<dataset:dataset>/badges/', endpoint='dataset_badges')
+class DatasetBadgesAPI(API):
+    @api.secure
```
Comment (noirbizarre) [inline udata/core/dataset/api.py]: Not needed because of following ``@api.secure(admin_permission)``

File: udata/core/dataset/api.py [hunk 3]
```diff
@@ -122,6 +126,39 @@ def delete(self, dataset):
...
+        badge = DatasetBadge(created=datetime.now(),
+                             created_by=current_user.id)
+        form.populate_obj(badge)
+        dataset.add_badge(badge)
+        return badge, 201
+
+
+@ns.route('/<dataset:dataset>/badges/<badge_kind>/', endpoint='dataset_badge')
+class DatasetBadgeAPI(API):
+    @api.secure
```
Comment (noirbizarre) [inline udata/core/dataset/api.py]: Not needed because of following ``@api.secure(admin_permission)``

File: udata/core/organization/api.py [hunk 4]
```diff
@@ -101,6 +107,39 @@ def delete(self, org):
         return '', 204
 
 
+@ns.route('/<org:org>/badges/', endpoint='organization_badges')
+class OrganizationBadgesAPI(API):
+    @api.secure
```
Comment (noirbizarre) [inline udata/core/organization/api.py]: Not needed because of following ``@api.secure(admin_permission)``

File: udata/core/organization/api.py [hunk 5]
```diff
@@ -101,6 +107,39 @@ def delete(self, org):
...
+        badge = OrganizationBadge(created=datetime.now(),
+                                  created_by=current_user.id)
+        form.populate_obj(badge)
+        org.add_badge(badge)
+        return badge, 201
+
+
+@ns.route('/<org:org>/badges/<badge_kind>/', endpoint='organization_badge')
+class OrganizationBadgeAPI(API):
+    @api.secure
```
Comment (noirbizarre) [inline udata/core/organization/api.py]: Not needed because of following ``@api.secure(admin_permission)``

File: udata/core/site/views.py [hunk 6]
```diff
@@ -169,13 +170,6 @@ def get_context(self):
                     'endpoint': 'organizations.list',
                 },
                 {
```
Comment (noirbizarre) [inline udata/core/site/views.py]: Don't we have an equivalent with badges ?

File: udata/core/badges/metrics.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+from .signals import on_badge_added, on_badge_removed
+
+
+class BadgesMetric(Metric):
+    name = 'badges'
+    display_name = _('Badges')
+
+    def get_value(self):
+        return (self.model_badge.objects(subject=self.target, removed=None)
```
Comment (noirbizarre) [inline udata/core/badges/metrics.py]: I think this is not working. It should be ``self.model.objects`` if I'm right

File: udata/core/site/views.py (hunk 6 above)
Comment (davidbgk) [inline udata/core/site/views.py]: I'm not sure that a metric for that is useful?

File: udata/core/badges/metrics.py (hunk 7 above)
Comment (davidbgk) [inline udata/core/badges/metrics.py]: Actually, `self.model` is the subject, see https://github.com/etalab/udata/pull/55/files#diff-8b1584d4d7d89f813bfbf43533ce51dfR50 for instance and `model_badge` is the inherited model from `Badge`.

File: udata/core/badges/models.py (hunk 1 above)
Comment (davidbgk) [inline udata/core/badges/models.py]: Given the display on the page, I think it's relevant to keep badges ordered by creation date. This way there will be "classic" ones first (public service, certified) and then additional and maybe more pertinent ones easier to scan for the visitor. I'm not sure about that though so maybe we should try and see once we've got more data?

File: udata/core/dataset/api.py (hunk 2 above)
Comment (davidbgk) [inline udata/core/dataset/api.py]: Thanks, copy-pasta evilness.

File: udata/core/badges/metrics.py (hunk 7 above)
Comment (noirbizarre) [inline udata/core/badges/metrics.py]: ok

File: udata/core/badges/models.py (hunk 1 above)
Comment (noirbizarre) [inline udata/core/badges/models.py]: Fine to me, seems more logical with the explanation :)

File: udata/core/site/views.py (hunk 6 above)
Comment (noirbizarre) [inline udata/core/site/views.py]: Not very much, but it's something that is used by the team to follow a little bit our "lobbying" progression.
Keep it like, we will see if someone ask for this.

**Comment by noirbizarre** [comment]
r+wc

File: udata/core/site/views.py (hunk 6 above)
Comment (davidbgk) [inline udata/core/site/views.py]: Given that we set those badges it's a bit biased ;)

Comment (noirbizarre) [inline udata/core/site/views.py]: It's only for keeping track of public services count ;)

**Pull request merged**